server/src/main/java/io/crate/execution/ddl/tables/TransportAddColumnAction.java

```diff
@@ -0,0 +1,231 @@
+                                    NodeContext nodeContext) {
+        super(ACTION_NAME,
+            transportService,
+            clusterService,
+            threadPool,
+            AddColumnRequest::new,
+            AcknowledgedResponse::new,
+            AcknowledgedResponse::new,
+            "add-column");
+        this.nodeContext = nodeContext;
+        this.indicesService = indicesService;
+    }
+
+    @Override
+    public ClusterStateTaskExecutor<AddColumnRequest> clusterStateTaskExecutor(AddColumnRequest request) {
+        return new DDLClusterStateTaskExecutor<>() {
+            @Override
+            protected ClusterState execute(ClusterState currentState, AddColumnRequest request) throws Exception {
+                Metadata.Builder metadataBuilder = Metadata.builder(currentState.metadata());
+                if (request.isPartitioned()) {
+                    String templateName = PartitionName.templateName(request.relationName().schema(), request.relationName().name());
+                    IndexTemplateMetadata indexTemplateMetadata = currentState.metadata().templates().get(templateName);
+
+                    Map<String, Object> existingTemplateMeta = new HashMap<>();
+                    mergeDeltaIntoExistingMapping(existingTemplateMeta, request);
+
+                    IndexTemplateMetadata newIndexTemplateMetadata = DDLClusterStateHelpers.updateTemplate(
+                        indexTemplateMetadata,
+                        existingTemplateMeta,
+                        Collections.emptyMap(),
+                        Settings.EMPTY,
+                        IndexScopedSettings.DEFAULT_SCOPED_SETTINGS // Not used if new settings are empty
+                    );
+                    metadataBuilder.put(newIndexTemplateMetadata);
+                }
+
+                currentState = updateMapping(currentState, request, metadataBuilder);
+                // ensure the new table can still be parsed into a DocTableInfo to avoid breaking the table.
+                new DocTableInfoFactory(nodeContext).create(request.relationName(), currentState);
+                return currentState;
+            }
+        };
+    }
+
+    /**
+     * Structure of the metadata is aligned with {@link AnalyzedTableElements#toMapping(AnalyzedTableElements)}
+     */
+    private ClusterState updateMapping(ClusterState currentState,
+                                       AddColumnRequest request,
+                                       Metadata.Builder metadataBuilder) throws IOException {
+        Index[] concreteIndices = resolveIndices(currentState, request.relationName().indexNameOrAlias());
+
+        for (Index index : concreteIndices) {
+            final IndexMetadata indexMetadata = currentState.metadata().getIndexSafe(index);
+
+            Map<String, Object> indexMapping = indexMetadata.mapping().sourceAsMap();
+            mergeDeltaIntoExistingMapping(indexMapping, request);
+
```

Comment:
I think before, `AlterTableClusterStateExecutor` eventually used `TransportPutMappingAction` which handled the negative positions.

`TransportAddColumnAction#updateMapping()` directly builds the cluster state so, the negative positions should be resolved before that.

```suggestion
TransportSchemaUpdateAction.populateColumnPositions(indexMapping);
```

Reply:
could you try adding the suggestion?

Reply:
Thanks, seems to be working, see https://github.com/crate/crate/pull/12806/commits/031be9ad940147545796916a3b23d92e61219ab1

>AlterTableClusterStateExecutor eventually used TransportPutMappingAction

not sure, I could see only `metadataMappingService.putMappingExecutor` which is the last executor updating cluster state.  I think `AlterTableClusterStateExecutor` worked as we did merge twice which I replicated in https://github.com/crate/crate/pull/12806/commits/e78c84ad4c967d5a4c8860199a3509c5648e8486. 

As far as I understand from debugging, extra merge(RECOVERY) makes non-master nodes enter [these](https://github.com/crate/crate/blob/master/server/src/main/java/org/elasticsearch/index/mapper/MapperService.java#L309-L310) lines which in turn triggers `ColumnPositionResolver`. Otherwise, `oldMapper` is null and no column position is triggered.